### PR TITLE
Implemented dprod

### LINKED ISSUE #3445: Implement Dprod
Implement Dprod Intrinsic Function
Refer: https://gcc.gnu.org/onlinedocs/gfortran/DPROD.html

### PR DESCRIPTION
**Pull request by parth121101**
Fix #3445

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -827,6 +828,37 @@ namespace Scale {
...
+        fill_func_arg("x", arg_types[0]);
+        fill_func_arg("y", arg_types[1]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /*
+        * r = dprod(x, y)
+        * r = x * y
+        */
+        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[0], ASR::cast_kindType::RealToReal, return_type, nullptr));
+        ASR::expr_t *cast1 = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[1], ASR::cast_kindType::RealToReal, return_type, nullptr));
+        body.push_back(al, b.Assignment(result, r_tMul(cast,cast1, return_type)));
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_functions.h:853]:
```suggestion
        body.push_back(al, b.Assignment(result, r64Mul(r2r64(args[0]), r2r64(args[1]))));
```

or 

```suggestion
        body.push_back(al, b.Assignment(result, r2r64(r32Mul(args[0],args[1]))));
```

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -827,6 +828,37 @@ namespace Scale {
     }
 }  // namespace Scale
 
+namespace Dprod {
+    static ASR::expr_t *eval_Dprod(Allocator &al, const Location &loc,
+            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_functions.h:833]:
```suggestion
            ASR::ttype_t* return_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -724,6 +724,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         {"floor", {IntrinsicSignature({"a", "kind"}, 1, 2)}},
         {"ceiling", {IntrinsicSignature({"a", "kind"}, 1, 2)}},
         {"scale", {IntrinsicSignature({"X", "I"}, 2, 2)}},
+        {"dprod", {IntrinsicSignature({"X", "Y"}, 2, 2)}},
```
Comment (Thirumalai-Shaktivel) [inline src/lfortran/semantics/ast_common_visitor.h:727]: Should we add this for all the functions? I don't think so.
We only add this for optional arguments or calling the functions with argument keywords.

Some examples, like `max`, we can have any number of args from 2 to 100. Or `maxloc(arr, back=true)` and so on.

**Review by certik — APPROVED** [review]
Didn't even know about this function.

If @Thirumalai-Shaktivel is happy, then we can merge this. It looks ok to me.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 3 above)
Comment (Thirumalai-Shaktivel) [inline src/lfortran/semantics/ast_common_visitor.h:727]:
```suggestion
```
Removing this cause any error?
If not, then we will remove it, as I said `name2signature` is only used by the functions having optional arguments.
Also, we have to remove `mod`, `hypot`, `shiftr`, ..., `scale`, ... from name2signature

**Comment by Thirumalai-Shaktivel** [comment]
Also, resolve the conflicts and this PR is good to merge!

File: src/lfortran/semantics/ast_common_visitor.h (hunk 3 above)
Comment (parth121101) [reply src/lfortran/semantics/ast_common_visitor.h:727]:
```
semantic error: No matching signature found for intrinsic dprod
lfortran/integration_tests/intrinsics_142.f90:8:14
     result = dprod(a,b)
                ^^^^^^^^^^ 
```
After removing this from ast_common_visitor.h file on running the integration tests this is what issue I am getting.

Comment (Thirumalai-Shaktivel) [reply src/lfortran/semantics/ast_common_visitor.h:728]: Okay, then it's fine. 
I think we need to update `get_intrinsic_signature` to handle these cases.

File: integration_tests/CMakeLists.txt [hunk 4]
```diff
@@ -650,6 +650,7 @@ RUN(NAME intrinsics_137 LABELS gfortran llvm)#huge
 RUN(NAME intrinsics_138 LABELS gfortran llvm)#precision
 RUN(NAME intrinsics_140 LABELS gfortran llvm)#selected_int_kind
 RUN(NAME intrinsics_141 LABELS gfortran llvm)#scale
+RUN(NAME intrinsics_142 LABELS gfortran llvm)#dprod
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/CMakeLists.txt:653]: Create a new file named `intrinsics_146` and add the tests there.

Comment (parth121101) [reply integration_tests/CMakeLists.txt:653]: Sure @Thirumalai-Shaktivel

File: integration_tests/CMakeLists.txt [hunk 5]
```diff
@@ -654,6 +654,7 @@ RUN(NAME intrinsics_142 LABELS gfortran llvm)#dot_product
 RUN(NAME intrinsics_143 LABELS gfortran llvm)#dot_product
 RUN(NAME intrinsics_144 LABELS gfortran llvm)#dot_product
 RUN(NAME intrinsics_145 LABELS gfortran llvm)#dot_product
+RUN(NAME intrinsics_146 LABELS gfortran llvm)#dprod
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/CMakeLists.txt:657]: Minor
```suggestion
RUN(NAME intrinsics_146 LABELS gfortran llvm) # dprod
```

**Review by Thirumalai-Shaktivel — APPROVED** [review]

**Pull request merged**